src/cargo/util/config/mod.rs

```diff
@@ -1643,7 +1643,59 @@ impl Config {
     /// Looks for a path for `tool` in an environment variable or config path, defaulting to `tool`
     /// as a path.
     fn get_tool(&self, tool: &str, from_config: &Option<ConfigRelativePath>) -> PathBuf {
+        // This function is designed to only work with rustup proxies. This
+        // assert is to ensure that if it is ever used for something else in
+        // the future that you must ensure that it is a proxy-able tool, or if
+        // not then you need to use `maybe_get_tool` instead.
+        assert!(matches!(tool, "rustc" | "rustdoc"));
         self.maybe_get_tool(tool, from_config)
+            .or_else(|| {
+                // This is an optimization to circumvent the rustup proxies
+                // which can have a significant performance hit. The goal here
+                // is to determine if calling `rustc` from PATH would end up
+                // calling the proxies.
+                //
+                // This is somewhat cautious trying to determine if it is safe
+                // to circumvent rustup, because there are some situations
+                // where users may do things like modify PATH, call cargo
+                // directly, use a custom rustup toolchain link without a
+                // cargo executable, etc. However, there is still some risk
+                // this may make the wrong decision in unusual circumstances.
+                //
+                // First, we must be running under rustup in the first place.
+                let toolchain = self.get_env_os("RUSTUP_TOOLCHAIN")?;
+                // This currently does not support toolchain paths.
+                // This also enforces UTF-8.
+                if toolchain.to_str()?.contains(&['/', '\\']) {
+                    return None;
+                }
+                // If the tool on PATH is the same as `rustup` on path, then
+                // there is pretty good evidence that it will be a proxy.
+                let tool_resolved = paths::resolve_executable(Path::new(tool)).ok()?;
+                let rustup_resolved = paths::resolve_executable(Path::new("rustup")).ok()?;
+                let tool_meta = tool_resolved.metadata().ok()?;
+                let rustup_meta = rustup_resolved.metadata().ok()?;
+                // This works on the assumption that rustup and its proxies
+                // use hard links to a single binary. If rustup ever changes
+                // that setup, then I think the worst consequence is that this
+                // optimization will not work, and it will take the slow path.
+                if tool_meta.len() != rustup_meta.len() {
+                    return None;
+                }
+                // Try to find the tool in rustup's toolchain directory.
+                let tool_exe = Path::new(tool).with_extension(env::consts::EXE_EXTENSION);
+                let toolchain_exe = home::rustup_home()
```

Comment:
`home` crate still access `RUSTUP_HOME` via `std::env`. Will there be an inconsistency when people set their `RUSTUP_HOME` in `[env]`?

https://github.com/rust-lang/cargo/blob/39684ff0aa88fe856460bc8665fd6f649b065e94/crates/home/src/env.rs#L93

Reply:
It shouldn't under normal circumstances, since the rustup proxies set `RUSTUP_HOME`, the `[env]` value will be ignored.

Reply:
I think it would be a good idea to prohibit this before stablisation @ehuss

Reply:
Sure, I went ahead and posted #12101.